internal/sync/lifecycle.go

```diff
@@ -117,39 +101,45 @@ func Once() LifecycleOnce {
 // If Start is called multiple times it will return the error
 // from the first time it was called.
 func (l *lifecycleOnce) Start(f func() error) error {
-	if l.starting.Swap(true) {
-		<-l.startCh
+	if l.state.CAS(int32(Idle), int32(Starting)) {
+		if f != nil {
+			l.err = f()
+		}
+
+		// skip forward to error state
+		if l.err != nil {
+			l.state.Store(int32(Errored))
+			close(l.stopCh)
+		} else {
+			l.state.Store(int32(Running))
+		}
+		close(l.startCh)
 		return l.err
 	}
-	if f != nil {
-		l.err = f()
-	}
-	// skip forward to error state
-	if l.err != nil {
-		l.errored.Store(true)
-		l.stopped.Store(true)
-		l.stopping.Store(true)
-		close(l.stopCh)
-	}
-	l.started.Store(true)
-	close(l.startCh)
 
+	<-l.startCh
 	return l.err
 }
 
 func (l *lifecycleOnce) WhenRunning(ctx context.Context) error {
-	if !l.stopping.Load() && l.started.Load() {
+	state := LifecycleState(l.state.Load())
+	if state == Running {
 		return nil
 	}
+	if state > Running {
+		return context.DeadlineExceeded
```

Comment:
I really have trouble with returning a context.DeadlineExceeded here. The deadline might not have been exceeded. The lifecycle could have be stopped right after starting. Shouldn't we return a more specific error for the sake of clarity?